### PR TITLE
UDAs for function arguments

### PR DESCRIPTION
**Pull request by IgorStepanov**
Implementation of subj.

This topic was partially discussed ~year ago in [NG](http://forum.dlang.org/post/pepqfcalhuonfszpmooc@forum.dlang.org).
The general idea:

```
class MyController
{
     View foo(@postParam("a") int p_a, @getParam("b") int p_b);
}

....
static if (is(typeof(MyController.foo) PAR == __parameters))
{
   enum attr_0 = __traits(getAttributes, PAR[0 .. 1]);
   enum attr_1 = __traits(getAttributes, PAR[1 .. 2]);
   static assert(attr_0 == postParam("a"));
   static assert(attr_1 == getParam("b"));
}
```

This feature is very usefull for different bindings generators like Web MVC or REST bindings like [Java Spring MVC](http://docs.spring.io/spring/docs/current/spring-framework-reference/html/mvc.html#mvc-ann-requestmapping).

### CODE REVIEW

File: test/runnable/uda.d [hunk 1]
```diff
@@ -467,6 +467,20 @@ static assert(__traits(getAttributes, FileData11844.member)[0](new FileData11844
 
 /************************************************/
 
+void test13x(@(10) int a, @(20) int, @(30) int[] arr...) {}
+
+void test13()
+{
+    static if (is(typeof(test12x) PT == __parameters))
```
Comment (schuetzm) [inline test/runnable/uda.d]: `test12x` vs `test13x`?

**Comment by schuetzm** [comment]
I'm afraid I don't understand that `static if()` construction. `typeof(MyController.foo)` is probably something like `View delegate(int, int)`. We would that ever be equal to `__parameters`?

Urgh... After some searching I found this:
http://forum.dlang.org/post/mailman.1197.1379014886.1719.digitalmars-d-learn@puremagic.com
I'm horrified :scream:

File: test/runnable/uda.d (hunk 1 above)
Comment (IgorStepanov) [inline test/runnable/uda.d]: Omg, this test was totally wrong. Fixed.

**Comment by IgorStepanov** [comment]
>I'm afraid I don't understand that static if() construction. typeof(MyController.foo) is probably something like View delegate(int, int). We would that ever be equal to __parameters?
Urgh... After some searching I found this:
http://forum.dlang.org/post/mailman.1197.1379014886.1719.digitalmars-d-learn@puremagic.com
I'm horrified :scream:

Yes, `__parameters` result looks strange.

You should write `__traits(identifier, PT[0 .. 1])` to get correct result. If you write `__traits(identifier, PT[0])` you will get a `__traits(identifier, int)` result.
See [phobos ](https://github.com/D-Programming-Language/phobos/blob/master/std/traits.d#L1118) soultion for more info.
Anywhere, this feature should be wrapped into high-level phobos template like `getArgumentUDAs!(func, <arg num>)`

**Comment by Geod24** [comment]
Having to emulate UDAs on parameters is doable, although it's a pain.

I'd really like to have something in the language to support it (as long as it doesn't conflict with other features / grammar), but `__parameters` is an horrible hack, and so are some of the things built on top of it, such as `ParameterDefaultValueTuple` which is at the moment totally broken for meta programming purposes.

Having a magical tuple that can hold things not expressible in any other part of the language (and that vanishes as soon as you put them through a tuple) is already bad enough, please, don't put more magic in it.

**Comment by MetaLang** [comment]
I think it's important to note that D can currently introduce new storage classes without breaking any code, due to the fact that UDAs on parameters are not allowed. This pull request would make that no longer true. For example, we could add a new storage class today and be confident that no code would be broken:

    struct Test { /*...*/ }

    void acceptRvals(@rval ref Test t); //@rval currently doesn't conflict with UDAs

Although I doubt that is much of a concern at this point.

**Comment by IgorStepanov** [comment]
>Having to emulate UDAs on parameters is doable, although it's a pain.

Yes, current `__parameters` interface looks like hack. However it is used for identifier or default value access now, and may be used for UDA access via a special Phobos template. 
This PR is about UDA, not about `__parameters`. If another way to access to parameters data will be introduced in the language, this UDA will work through it.

>I think it's important to note that D can currently introduce new storage classes without breaking any code, due to the fact that UDAs on parameters are not allowed.

The same point for global UDAs. However, when `@nogc` was being implemented, UDAs was exist and potentially may cause a conflict with the new attribute.

**Comment by 9rnsr** [comment]
@IgorStepanov Additionally you need to tweak `stripDefaultArgs` in `mtype.c`. Because:

```d
enum uda;
void foo(@uda int v) {}

alias x = X!(typeof(foo));

template X(F) {
    // in here, the function type F must not have any udas.
}
```

**Comment by IgorStepanov** [comment]
@9rnsr Done. Also I've added the corresponding test.
@CyberShadow  What is the trouble with doc tester? How can I fix it?

**Comment by CyberShadow** [comment]
> @CyberShadow What is the trouble with doc tester? How can I fix it?

Git repository corruption I think. Fixed for now, hopefully it won't recur

**Comment by IgorStepanov** [comment]
@9rnsr Any other objections?
>Git repository corruption I think. Fixed for now, hopefully it won't recur

Thanks.

File: src/parse.c [hunk 2]
```diff
@@ -1940,7 +1942,27 @@ Parameters *Parser::parseParameters(int *pvarargs, TemplateParameters **tpl)
...
                     goto L2;
+                case TOKat:
+                {
+                    Expressions *exps = NULL;
+                    stc = parseAttribute(&exps);
+
+                    if (stc == STCproperty || stc == STCnogc || stc == STCdisable ||
+                        stc == STCsafe || stc == STCtrusted || stc == STCsystem)
+                    {
+                        error("@%s attribute for function argument is not supported", token.toChars());
```
Comment (9rnsr) [inline src/parse.c]: Better wording: s/argument/parameter/

File: src/parse.c [hunk 3]
```diff
@@ -1940,7 +1942,27 @@ Parameters *Parser::parseParameters(int *pvarargs, TemplateParameters **tpl)
...
+                    }
+                    else
+                    {
+                        udas = UserAttributeDeclaration::concat(udas, exps);
+                    }
 
+                    if (token.value == TOKdotdotdot)
+                        error("variadic argument cannot has UDAs");
```
Comment (9rnsr) [inline src/parse.c]: Better wording: s/argument/parameter/

**Comment by 9rnsr** [comment]
Excepting nitpicks for the diagnostic messages, the implementation looks good to me.

File: src/parse.c (hunk 2 above)
Comment (IgorStepanov) [inline src/parse.c]: fixed

File: src/parse.c (hunk 3 above)
Comment (IgorStepanov) [inline src/parse.c]: fixed

**Comment by IgorStepanov** [comment]
>Excepting nitpicks for the diagnostic messages, the implementation looks good to me.

I've fixed messages. Thanks!

**Comment by Geod24** [comment]
> This PR is about UDA, not about __parameters. If another way to access to parameters data will be introduced in the language, this UDA will work through it.

My point was that as soon as it's introduced, we'll have to support it (unless it doesn't make it into a release). But since this would ease my life a lot, I won't complain too much :-)

Is there a corresponding documentation change open ?

Also, any chance you can add a test for the following scenarios:
- C variadic parameters;
- UDAs on the RHS of the identifier (fails with a meaningfull error message);
- Taking aggregates / using a function return value (like in Vibe.d);
- Types UDAs (including one with an `enum Foo;`);
- UDAs on `function` / `delegate` parameters (including ones with `@nogc` or `@safe`);

**Comment by IgorStepanov** [comment]
@Geod24 
>My point was that as soon as it's introduced, we'll have to support it (unless it doesn't make it into a release).

We will use phobos wrapper template in high-level code. Similar to `ParameterIdentifierTuple` and other. See the simple example of this template in the test `runnable/uda.d`.
I am also not happy with the current situation with `__parameters`, but if we start changing this situation now, we will finish this work near Christmas. Maybe. :)

>Also, any chance you can add a test for the following scenarios:

Yes, done.

>C variadic parameters;
>UDAs on the RHS of the identifier (fails with a meaningfull error message);

See `fail_compilation/udaparams.d`

>Taking aggregates / using a function return value (like in Vibe.d);
>Types UDAs (including one with an enum Foo;);
>UDAs on function / delegate parameters (including ones with @nogc or @safe);

See `runnable/uda.d`

Any other ideas?

**Comment by IgorStepanov** [comment]
@9rnsr I've changed `parse.c` and `func.c` code. (Added error for postfix uda declaration and passing uda for parameter `VarDeclaration*` in `FuncDeclaration`.
Is it ok?

**Comment by IgorStepanov** [comment]
@9rnsr Please, revalidate this. I've added two changes, declared in previous post.

File: src/parse.c [hunk 4]
```diff
@@ -1940,7 +1942,27 @@ Parameters *Parser::parseParameters(int *pvarargs, TemplateParameters **tpl)
...
+                    {
+                        error("@%s attribute for function parameter is not supported", token.toChars());
+                    }
+                    else
+                    {
+                        udas = UserAttributeDeclaration::concat(udas, exps);
+                    }
 
+                    if (token.value == TOKdotdotdot)
+                        error("variadic parameter cannot has UDAs");
```
Comment (9rnsr) [inline src/parse.c]: s/cannot has/cannot have/

**Comment by 9rnsr** [comment]
Looks good excepting one diagnostic nitpick.

File: test/fail_compilation/udaparams.d [hunk 5]
```diff
@@ -0,0 +1,24 @@
+/*
+TEST_OUTPUT:
+---
+fail_compilation/udaparams.d(15): Error: variadic parameter cannot has UDAs
+fail_compilation/udaparams.d(16): Error: variadic parameter cannot has UDAs
```
Comment (MetaLang) [inline test/fail_compilation/udaparams.d]: This is inconsistent with the error message on the very next line, which uses `user defined attributes` as opposed to `UDAs`. You should use either one or the other, but not both.

Comment (IgorStepanov) [inline test/fail_compilation/udaparams.d]: Ok, will change to "user defined attributes".

**Comment by TheFlyingFiddle** [comment]
What's the status on this?

**Comment by wilzbach** [comment]
Revived in #7576

**Pull request closed without merging**